pkg/services/accesscontrol/resourcepermissions/service.go

```diff
@@ -450,21 +453,126 @@ type ActionSet struct {
 	Actions []string `json:"actions"`
 }
 
-// InMemoryActionSets is an in-memory implementation of the ActionSetService.
-type InMemoryActionSets struct {
-	features           featuremgmt.FeatureToggles
-	log                log.Logger
-	actionSetToActions map[string][]string
-	actionToActionSets map[string][]string
+type ActionSetStore interface {
+	// StoreActionSet stores action set. If a set with the given name has already been stored, the new actions will be appended to the existing actions.
+	StoreActionSet(name string, actions []string)
+	// ResolveActionSet resolves an action set to a list of corresponding actions.
+	ResolveActionSet(actionSet string) []string
+	// ResolveAction returns all the action sets that the action belongs to.
+	ResolveAction(action string) []string
+	// ResolveActionPrefix returns all action sets that include at least one action with the specified prefix
+	ResolveActionPrefix(prefix string) []string
+	// ExpandActionSetsWithFilter takes a set of permissions that might include some action set permissions, and returns a set of permissions with action sets expanded into underlying permissions.
+	// When action sets are expanded into the underlying permissions only those permissions whose action is matched by actionMatcher are included.
+	ExpandActionSetsWithFilter(permissions []accesscontrol.Permission, actionMatcher func(action string) bool) []accesscontrol.Permission
+}
+
+type ActionSetSvc struct {
+	features featuremgmt.FeatureToggles
+	store    ActionSetStore
 }
 
 // NewActionSetService returns a new instance of InMemoryActionSetService.
 func NewActionSetService(features featuremgmt.FeatureToggles) ActionSetService {
-	actionSets := &InMemoryActionSets{
-		features:           features,
-		log:                log.New("resourcepermissions.actionsets"),
-		actionSetToActions: make(map[string][]string),
-		actionToActionSets: make(map[string][]string),
+	return &ActionSetSvc{
+		features: features,
+		store:    NewInMemoryActionSetStore(features),
+	}
+}
```

Comment:
this is just me but, should we have a logger here as well. that is being used for the store as well.

```suggestion
	return &ActionSetSvc{
		features: features,
		log: log.New("actionsetlogger"),
		store:    NewInMemoryActionSetStore(features),
	}
}
```

Reply:
I think we can add it to the service if we start using it in the service, but currently it feels unnecessary imo.